sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py

```diff
@@ -0,0 +1,224 @@
+    one_incr = np.int32(1)
+    one_decr = np.int32(-1)
+    zero_init = dtype(0.0)
+    max_candidate_id = np.int32(n_samples - 1)
+
+    @dpex.kernel
+    # fmt: off
+    def sample_center_candidates(
+        closest_dist_sq,          # IN             (n_features, n_samples)
+        total_potential,          # IN             (1,)
+        random_state,             # INOUT          (n_local_trials, 2)
+        candidates_id,            # OUT            (n_local_trials,)
+        ):
+    # fmt: on
+        local_trial_idx = dpex.get_global_id(zero_idx)
+        if local_trial_idx >= n_local_trials:
+            return
+        random_value = (rand_uniform_kernel_func(random_state, local_trial_idx)
+                        * total_potential[zero_idx])
+
+        cumulative_potential = zero_init
+        candidate_id = one_decr
+        while (random_value > cumulative_potential) and (candidate_id < max_candidate_id):
+            candidate_id += one_incr
+            cumulative_potential += closest_dist_sq[candidate_id]
+        candidates_id[local_trial_idx] = candidate_id
+
+    global_size = (math.ceil(n_local_trials / work_group_size)) * (work_group_size)
+    return sample_center_candidates[global_size, work_group_size]
+
+
+@lru_cache
+def make_kmeansplusplus_single_step_fixed_window_kernel(
+    n_samples,
+    n_features,
+    n_candidates,
+    preferred_work_group_size_multiple,
+    centroids_window_width_multiplier,
+    centroids_window_height,
+    work_group_size,
+    dtype,
+):
+
+    window_n_centroids = (
+        preferred_work_group_size_multiple * centroids_window_width_multiplier
+    )
+
+    (
+        _initialize_window_of_centroids,
+        _load_window_of_centroids_and_features,
+    ) = _make_initialize_window_kernel_funcs(
+        n_samples,
+        n_features,
+        work_group_size,
+        window_n_centroids,
+        centroids_window_height,
+        dtype,
+    )
+
+    _accumulate_sq_distances = _make_accumulate_sum_of_ops_kernel_func(
+        n_samples,
+        n_features,
+        centroids_window_height,
+        window_n_centroids,
+        ops="squared_diff",
+        dtype=dtype,
+    )
+
+    n_windows_per_feature = math.ceil(n_candidates / window_n_centroids)
+    n_windows_per_centroid = math.ceil(n_features / centroids_window_height)
+
+    centroids_window_shape = (centroids_window_height, (window_n_centroids + 1))
+
+    zero_idx = np.int64(0)
+
+    @dpex.kernel
+    # fmt: off
+    def kmeansplusplus_single_step(
+        X_t,                               # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,                     # IN READ-ONLY   (n_samples,)
+        candidates_ids,                    # IN             (n_candidates,)
+        closest_dist_sq,                   # IN             (n_samples,)
+        euclidean_distances_t,             # OUT            (n_candidates, n_samples)
+    ):
+    # fmt: on
+        sample_idx = dpex.get_global_id(zero_idx)
```

Comment:
Since we parallelize on `sample_idx`, I would expect this code to be the fastest when `X_t` has a Fortran layout (contiguous features for a given sample).

But at the moment we raise `ValueError` instead of allocating `X_t` with Fortran layout (that would correspond to C layout for an on-device `X` that does not exist) in `_load_transposed_data_to_device`. But I don't understand why. Both layouts seems to work fine in my setting with dpctl.__version__ == "0.13.0" from today's docker container:

```python
>>> import numpy as np
>>> import dpctl
>>> dpctl.__version__
'0.13.0'
>>> gpu_device = dpctl.SyclDevice("gpu")
<dpctl.SyclDevice [backend_type.level_zero, device_type.gpu,  Intel(R) Iris(R) Graphics 540 [0x1926]] at 0x7ff09594b370>

>>> data_np = np.arange(12).reshape(3, 4)
>>> data_np.T
array([[ 0,  4,  8],
       [ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11]])

>>> data_t_c = dpctl.tensor.from_numpy(data_np.T, device=gpu_device)
>>> data_t_c.strides
(3, 1)  # C layout for `data_t_c`
>>> dpctl.tensor.asnumpy(data_t_c)
array([[ 0,  4,  8],
       [ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11]])

>>> data_t_f = dpctl.tensor.from_numpy(data_np, device=gpu_device).T
>>> data_t_f.strides
(1, 4)  # F layout for data_t_f
>>> dpctl.tensor.asnumpy(data_t_f)
array([[ 0,  4,  8],
       [ 1,  5,  9],
       [ 2,  6, 10],
       [ 3,  7, 11]])
```

So why don't we work with `X_t` allocated on-device with a Fortran layout?

Reply:
`numba_dpex` [does not](https://github.com/IntelPython/numba-dpex/issues/767) support inputs with `F-layout` so it's hard to try our kernels where different layouts, it would require a hack that consists in passing `X` or `X_t` in `C-layout` and inverting the arguments of all `getitem` operations in the kernels depending on wether `X` or `X_t` is passed (perhaps could sound easier if mentionning that `getitem`s could be moved to abstracted kernels that take the layout as instantiation parameter but it would be very verbose for just `getitem`s)

From the information I've gathered before starting the project, our gpu kernels and especially pairwise distance-like kernels should perform better when `X` is `F-layout` so that's why we use `X_t` everywhere.

For the sampling step we could use X in C-Layout if it's more efficient but if changing layout triggers a copy (does it?) it wouldn't be worth it, maybe a compute overhead would be acceptable.

Reply:
I opened to clarify https://github.com/soda-inria/sklearn-numba-dpex/pull/53 this.